evm/p2p/les.py

```diff
@@ -6,14 +18,159 @@
 
 class Status(Command):
     _cmd_id = 0
+    decode_strict = False
+    # A list of (key, value) pairs is all a Status msg contains, but since the values can be of
+    # any type, we need to use the raw sedes here and do the actual deserialization in
+    # decode_payload().
+    structure = sedes.CountableList(sedes.List([sedes.binary, sedes.raw]))
+    # The sedes used for each key in the list above.
+    items_sedes = {
+        'protocolVersion': sedes.big_endian_int,
+        'networkId': sedes.big_endian_int,
+        'headTd': sedes.big_endian_int,
+        'headHash': sedes.binary,
+        'headNum': sedes.big_endian_int,
+        'genesisHash': sedes.binary,
+        'serveHeaders': None,
+        'serveChainSince': sedes.big_endian_int,
+        'serveStateSince': sedes.big_endian_int,
+        'txRelay': None,
+        'flowControl/BL': sedes.big_endian_int,
+        'flowControl/MRC': sedes.CountableList(
+            sedes.List([sedes.big_endian_int, sedes.big_endian_int, sedes.big_endian_int])),
+        'flowControl/MRR': sedes.big_endian_int,
+    }
+
+    def decode_payload(self, rlp_data):
+        data = super(Status, self).decode_payload(rlp_data)
+        # The LES/Status msg contains an arbitrary list of (key, value) pairs, where values can
+        # have different types and unknown keys should be ignored for forward compatibility
+        # reasons, so here we need an extra pass to deserialize each of the key/value pairs we
+        # know about.
+        decoded = {}
+        for key, value in data:
+            # The sedes.binary we use in .structure above will give us a bytes value here, but
+            # using bytes as dictionary keys makes it impossible to use the dict() constructor
```

Comment:
I'm not following what exactly is broken here?

```python
In [1]: x = {}

In [2]: x[b'arst'] = 3

In [3]: x
Out[3]: {b'arst': 3}

In [4]: x = {b'arst': 3}

In [5]: x
Out[5]: {b'arst': 3}
```

Seems to work as expected to me?  I must be missing something.

Reply:
If I construct the dictionary using dict() and keyword arguments, then later use the values from the decoded payload, it won't work because the dictionary will have strings as keys whereas the decoded payload will be bytes. I did this just to standardise on strings as dict keys, as that's what I used in .items_sedes as well

```python
In [1]: d = dict(foo='bar')

In [2]: d[b'foo']
------- KeyError                                  Traceback (most recent call last) 
<ipython-input-2-d73dd6c877f2> in <module>()
----> 1 d[b'foo']

KeyError: b'foo'
```